### DISCUSSION TITLE
creation of poly region

### DISCUSSION THREAD

**Discussion by Deba57** [dtcenter/METplus#2155] [Configuration] [answered]
Hi,
How to create mask file over a continent using gen-vx tool in MET ? Input is global data and want to create a region only over Australia as observations are over that part only.

Thanks
Debasish

**Comment by JohnHalleyGotway** [comment]
Hi Deb, I see you'd like to use the gen_vx_mask tool to create a masking region over Australia for a global model domain. I'd recommend using the "[-type shape](https://met.readthedocs.io/en/latest/Users_Guide/masking.html#optional-arguments-for-gen-vx-mask)" shapefile masking option to do this.

1. Go to the natural earth data [downloads page](https://www.naturalearthdata.com/downloads/).
2. Download the large scale (1:110m) [country data](https://www.naturalearthdata.com/http//www.naturalearthdata.com/download/110m/cultural/ne_110m_admin_0_countries.zip).
3. Run MET's `gis_dump_dbf` utility to figure out which record you want to extract:
```
gis_dump_dbf ne_110m_admin_0_countries/ne_110m_admin_0_countries.dbf | egrep -i "Record |Australia"
```
Note that all the info about Australia is in record number 137:
```
Record 137 ...
|    SOVEREIGNT = "Australia"
|    ADMIN      = "Australia"
...
```
4. Run `gen_vx_mask` to define a mask of global 1/2 degree NCEP lat/lon grid "[G004](https://www.nco.ncep.noaa.gov/pmb/docs/on388/grids/grid004.gif)" for shape number 137:
```
gen_vx_mask G004 ne_110m_admin_0_countries/ne_110m_admin_0_countries.shp G004_Australia.nc -type shape -shapeno 137 -name Australia -v 3
```
Here's the log output:
```
DEBUG 1: Input Grid:		G004
DEBUG 1: Mask File:		ne_110m_admin_0_countries/ne_110m_admin_0_countries.shp
DEBUG 3: Use input grid named "G004".
DEBUG 2: Parsed Input Grid:	G004 (720 x 361)
DEBUG 2: Parsed Shape Mask:	ne_110m_admin_0_countries/ne_110m_admin_0_countries.shp containing 241 points
DEBUG 3: Shape Masking:		2783 of 259920 points inside
DEBUG 1: Output File:		G004_Australia.nc
```
5. Run `plot_data_plane` to confirm the result:
```
plot_data_plane G004_Australia.nc G004_Australia.ps 'name="Australia";level="(*,*)";'
```
Here's the resulting image:
![Screen Shot 2023-05-01 at 4 30 27 PM](https://user-images.githubusercontent.com/21087144/235542801-813f0c1a-df03-4eee-b651-ee06a2e514d9.png)

Of course you can replace "G004" with some other named NCEP grid... or just provide the path to a gridded data file that MET knows how to read, and it'll use that input grid instead.

Hope this helps.

Thanks,
John

**Reply by Deba57** [reply]
This is great @JohnHalleyGotway. Thank you very much.
One quick question, same source of data to be used to create similar mask files over continents like Europe or Asia ? 

Deba.

**Reply by JohnHalleyGotway** [reply, marked as answer]
@Deba57 that is a good question. It just so happens that Australia is both a country AND a continent. So that's the easy one!

Europe is more difficult one since it consists of many countries. The Natural Earth Data shapefile DOES indicate the continent for each country but you have to stitch those countries together into the continent.

For example:
```
bin/gis_dump_dbf ne_110m_admin_0_countries/ne_110m_admin_0_countries.dbf | egrep "Record| NAME |CONTINENT"
```
Dumps out:
```
...
Record 142 ...
|    NAME       = "Denmark"
|    CONTINENT  = "Europe"
...
```
And here's a very long `gis_dump_dbf` command to figure out which records belong to Europe:
```
gis_dump_dbf ne_110m_admin_0_countries/ne_110m_admin_0_countries.dbf | egrep -i "^Record |Continent" | egrep -B 1 -a Europe | egrep -a Record | cut -d' ' -f2 | sed -r 's/$/ /g' | tr -d '\n'
```
Here's the 39 records that comprise Europe:
```
18 21 43 110 111 112 113 114 115 116 117 118 119 120 121 122 123 125 126 127 128 129 130 131 132 133 141 142 143 144 150 151 152 153 170 171 172 173 174
```
Ideally, you'd be able to list all 39 record numbers to use all the shapes. Or better yet, tell gen_vx_mask to use any record where `CONTINENT = Europe`. Unfortunately though that doesn't exist. As of now, only a single one can be specified. So instead, you need to run gen_vx_mask 39 times... once for each record.
```
# Run once to get started with a field of all 0's.
# Shifted to keep the mask in the center of the domain and avoid wrap effects.
bin/gen_vx_mask \
'latlon 720 361 -90 -130 0.5 0.5' \
'latlon 720 361 -90 -130 0.5 0.5' \
-type grid -complement europe_mask_in.nc 

# Run once for each country to change those 0's to 1's
for n in `echo "18 21 43 110 111 112 113 114 115 116 117 118 119 120 121 122 123 125 126 127 128 129 130 131 132 133 141 142 143 144 
150 151 152 153 170 171 172 173 174"`; do
  bin/gen_vx_mask \
    europe_mask_in.nc \
    ne_110m_admin_0_countries/ne_110m_admin_0_countries.shp \
    europe_mask.nc \
    -type shape -shapeno ${n} -name Europe -value ${n}
  cp europe_mask.nc europe_mask_in.nc
done
```
And then plot the result.
```
bin/plot_data_plane europe_mask.nc europe_mask.ps 'name="Europe"; level="(*,*)";'
```
Here's the image:
<img width="609" alt="Screen Shot 2023-05-02 at 10 45 59 AM" src="https://user-images.githubusercontent.com/21087144/235731233-bb2fb98b-c6d8-4d76-9bcf-6db8aa0d6dbe.png">

I used `-value ${n}` to write the Record number instead of just 1. But removing that argument would just make it a 1.

So there's clearly some issues with the result. We shouldn't include ALL of Russia and what's going on in South America.

So this just demonstrates how `gen_vx_mask` can be run iteratively to combine multiple masks. But obviously it'd be preferable to enhance it to process multiple shapes in a single run.
